### ISSUE TITLE
Phase 2: ML × formula blend for race-day prediction (vLT × distance penalty, eFTP source)

### ISSUE THREAD

**Issue by radikkhaziev** [radikkhaziev/triathlon-agent#362]
## Context

Root fix for #359 Q1 (run prediction barely reacts to projected CTL). Phase 1.7 fixed CI envelope but didn't touch prediction math — XGBoost trees still clip on out-of-sample CTL, so race_day pred ≈ today pred (4 sec/km delta despite CTL projected 30 → 66).

Phase 2.0 blends ML with a transparent physiology formula. Weight transitions: far from race → formula leads, close to race → ML leads. Spec lives in §10.5.

## Approach

```python
ml_pred      = xgboost.predict(...)                        # current — leaf-clipped
formula_pred = vlt_distance_penalty(eFTP_run, race_distance_m, race_type)
weight       = schedule(days_to_race)                      # see 2.0a — empirical, not linear
final_pred   = ml_pred * weight + formula_pred * (1 - weight)
```

### Formula choice: vLT × distance penalty (not Riegel, not VDOT)

- **Riegel** rejected: works between same-type distances. 70.3 Run is structurally different from standalone HM (bike fatigue). Riegel lies here.
- **VDOT** rejected: requires polynomial fit on threshold_pace, adds a layer to validate. Not value-add for Phase 2.0.
- **vLT × penalty** chosen: transparent table, verifiable on athlete's own races history.

Distance penalty table (race-context aware — critical for triathlon):

| Race                          | Distance | Penalty (sec/km vs threshold) |
|-------------------------------|----------|-------------------------------|
| 5K standalone                 | 5000     | −5                            |
| 10K standalone                | 10000    | 0                             |
| Half marathon standalone      | 21097    | +5 to +15                     |
| **70.3 Run (after bike)**     | 21097    | **+30 to +45 (default +37)**  |
| Marathon standalone           | 42195    | +20 to +35                    |
| IM Run (after bike+long swim) | 42195    | +60 to +90                    |

Constants live in `data/ml/formula_constants.py` for retuning during 2.0c calibration.

### Threshold source: eFTP_run from FITNESS_UPDATED, not athlete_settings.threshold_pace

```python
projection = await FitnessProjection.get(user_id, race_date)
eftp_run = projection.sport_info_by_type("Run", "eftp")
if eftp_run is not None:
    threshold_pace_sec_per_km = 1000.0 / eftp_run        # m/s → sec/km
else:
    # Fallback 1: current eFTP from latest Wellness sport_info
    # Fallback 2: athlete_settings.threshold_pace (staleness-prone)
    # Fallback 3: refuse to blend (formula_weight = 0 → ML only)
```

Why: `athlete_settings.threshold_pace` only updates on drift detection after ramp test with R²≥0.85. Between tests → stale. Intervals daily-recomputes eFTP from rolling power-pace data, so it auto-tracks CTL changes, detraining, overreach.

**Stale guards (two layers):**
1. At train time: bundle stores `metrics.threshold_pace_age_days_at_train`
2. At predict time:
   - `now() - last_zones_update > 90d` AND fallback to athlete_settings → warning `threshold_pace_stale`, formula_weight reduced or refused
   - **NEW:** `(projected_ctl - current_ctl) / current_ctl > 0.5` → warning `threshold_pace_pre_ramp` (even if zones fresh, threshold may not reflect race-day fitness after major CTL ramp like 25→66)

### 2.0a — Validation (no production code)

**What:** `tools/race_blend_simulation.py` standalone script
1. For each race in `races` table (user 1 has n=22, of which 19 Run):
   - For each `days_out ∈ {30, 60, 90, 120, 150}`:
     - Snapshot features as of `(race_date - days_out)`
     - Compute `ml_pred` (loaded model)
     - Compute `formula_pred` (proposed table above)
     - Get `actual` race pace
2. Compute `optimal_weight(days_out) = argmin |w*ml + (1-w)*formula − actual|` per race
3. Plot avg `optimal_weight` vs `days_out` → reveals shape (linear / sigmoid / threshold)

**Cross-validate** on user 14 + 62 (each n≥10 races) — n=22 alone is too small to fit shape without overfitting to one athlete's patterns. Treat output as **directional hint, not production fit**.

**Effort:** ~80 LoC standalone, not shipped to prod.

**Expected output:**
- Plot weight(days_out) per athlete + averaged
- Variance per athlete — if huge, per-athlete tuning matters more than universal constants
- Residual sign — if ML systematically overestimates, blend is more valuable than expected
- Recommended weight schedule (hypothesis: sigmoid between ~0.2 at 180d and ~0.8 at race day, inflection ~60-90d)

### 2.0b — Implementation

**What:** Formula blend in `_predict_one` (or wherever run-pred lives), edge case tests, surfacing of new warnings.

**Effort:** ~150 LoC + 7 tests (see Test Plan).

**Ship gate:** 2.0a results approved by user.

### 2.0c — Calibration (ongoing)

**What:** Constants in `data/ml/formula_constants.py` stay tweakable. Retune after 3-6 months of accumulated race data.

**Effort:** TBD, doesn't block 2.0b ship.

## Test Plan (for 2.0b)

```python
class TestFormulaBlend:
    def test_ctl_drop_injury(self):
        """CTL=50 → crashed to 25 due to injury, race in 60d, plan CTL=70.
        Cap (#349 linear interp) engages at 50→50, formula gets capped CTL,
        blend = ML conservative + formula moderate → middle ground."""

    def test_ctl_ramp_faster_than_plan(self):
        """CTL=80 at race day, plan was 70. out_of_sample_ctl warning fires.
        Formula extrapolates honestly to high pace, blend = optimistic + warning."""

    def test_missing_eftp_falls_back_to_threshold_pace(self):
        """sport_info_by_type('Run', 'eftp') → None.
        Falls back to athlete_settings.threshold_pace, formula applies normally."""

    def test_no_threshold_pace_no_eftp_pure_ml(self):
        """Cold-start athlete, neither eFTP nor threshold_pace.
        formula_weight=0, ML-only behavior preserved, warning 'formula_unavailable_no_threshold'."""

    def test_stale_threshold_pace_reduces_formula_weight(self):
        """threshold_pace 180 days old, no eFTP.
        formula_weight × 0.5 OR refused entirely, warning 'threshold_pace_stale'."""

    def test_race_day_70_3_run_distance_penalty(self):
        """70.3 Run penalty (+30 to +45) > standalone HM (+5 to +15)
        even with same distance + threshold_pace."""

    def test_blend_monotonic_on_days_to_race(self):
        """Fixed CTL, varied days_out: weight schedule must be monotonic
        (no bumps from sigmoid edge cases)."""
```

Plus integration test: `predict_splits_with_ci` end-to-end with blend for all 3 modes (today, race_day fresh, race_day stale).

## Affected files

- `data/ml/formula_constants.py` — new, holds penalty table + weight schedule constants
- `tools/race_blend_simulation.py` — new, 2.0a only
- Wherever `_predict_one` / race-pred lives (likely under `models/race_projection/`)
- Test file alongside the predict module
- `docs/RACE_PROJECTION_SPEC.md` §10.5 — design fixation (do before 2.0a)

## Acceptance criteria

**Per sub-phase:**

- **2.0a:** Simulation script runs on user 1 + 14 + 62. Plot + numbers shared. User approves weight schedule shape before 2.0b starts.
- **2.0b:** All 7 unit tests + integration test pass. race_day Run pred reacts meaningfully to projected_ctl ramp (e.g., for user 1 at 126 days out with projected CTL 66 vs current 30, race-day pred shifts from current 5:51/km toward ~5:25-5:35/km range, not stuck at ML-clipped value).
- **2.0c:** No code criteria, marker only.

**Overall:**
- Closes #359 Q1
- New warnings (`threshold_pace_stale`, `threshold_pace_pre_ramp`, `formula_unavailable_no_threshold`) documented in tool response schema
- Spec §10.5 reflects final design

## Related

- Parent: #359 (Q1)
- Sibling: #361 (Q2 envelope metadata)
- Spec: `docs/RACE_PROJECTION_SPEC.md` §10.5
- Phase 1.7 (CI envelope cap + warnings) ships before this
- Distance penalty constants source: AthleticGenes / Strava data + Joe Friel's coaching tables
- Validation surfaced during chat session 2026-05-12

---
_Reported via MCP by user_id=1_

**Comment by radikkhaziev** [comment]
URL: https://github.com/radikkhaziev/triathlon-agent/issues/362

```markdown
Closing as **not planned**. Formula blend approach is below noise floor — superseded by #363 (post-hoc bias correction).

## Simulation results

Ran on user 1, n=22 races, bias-corrected formula across horizons {30, 60, 90, 120, 150} days_out:

| Strategy                          | MAE (sec/km) | Δ vs ML       |
|-----------------------------------|--------------|---------------|
| ML only                           | 55.04        | —             |
| Formula only (bias-corrected)     | 56.07        | +1.03 worse   |
| 50/50 blend                       | 54.43        | -0.61         |
| Linear blend (0.70 → 0.52)        | 54.35        | -0.69         |
| **Oracle blend (per-race optimal)** | **42.53**  | **-12.51**    |

## Per-spec §10.5.5 gate

| Criterion                            | Result                  | Reading      |
|--------------------------------------|-------------------------|--------------|
| Slope ≥ 0.2                          | 0.18                    | ⚠️ Marginal  |
| Statistical significance (z > 1.65)  | z = 1.13 (p ≈ 0.13)     | ❌ Failed    |
| MAE improvement (target ≥ 5 sec/km)  | 0.69 sec/km             | ❌ Failed    |
| Per-horizon dominance                | Marginal at any horizon | ❌ Failed    |

🔴 **RED.** Direction correct but magnitude indistinguishable from noise.

## Why it didn't work

Per-race variance dominates (σ ≈ 60 sec/km) vs horizon effect (≈ 16 sec/km). Formula alone is no better than ML at any horizon — both static lookups can't reduce per-race noise driven by trail/road, weather, fatigue context, etc.

Oracle MAE 42.5 vs ML 55.0 = real 12.5 sec/km headroom, but it lives in **race-specific features**, not horizon-based blending.

## Pivot tracker

→ **#363** — three sub-paths:
- **β2** (in progress, soon-shipping): post-hoc bias correction. MAE 55.04 → 50.04 sec/km, z=2.63, p<0.01. 🟢 GREEN. Closes #359 Q1.
- **β1** (deferred): race-feature enrichment (is_trail, weather, elevation_per_km). Targets the 12.5 sec/km Oracle headroom.
- **β3** (deferred): cross-athlete pool model. Requires multi-tenant race data.

Spec §10.5 will be rewritten under #363 to reflect this pivot.
```
